python/paddle/fluid/dygraph/dygraph_to_static/ast_utils.py

```diff
@@ -0,0 +1,323 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+
+import astor
+import gast
+import six
+import copy
+import tempfile
+import imp
+import os
+import atexit
+from collections import defaultdict
+
+from paddle.fluid import unique_name
+
+TRUE_FUNC_PRFIX = 'true_fn'
+FALSE_FUNC_PRFIX = 'false_fn'
+
+
+def is_control_flow_if(node):
+    """
+    Determine whether the node is a plain python `if statement` or
+    control flow in Paddle.
+    """
+    return True
+
+
+def get_name_ids(nodes, not_name_set=None, node_black_list=None):
+    """
+    Return all ast.Name.id of python variable in nodes.
+    """
+    if not isinstance(nodes, (list, tuple, set)):
+        raise ValueError(
+            "nodes must be one of list, tuple, set, but received %s" %
+            type(nodes))
+    if not_name_set is None:
+        not_name_set = set()
+
+    def update(old_dict, new_dict):
+        for k, v in new_dict.items():
+            old_dict[k].extend(v)
+
+    name_ids = defaultdict(list)
+    for node in nodes:
+        if node_black_list and node in node_black_list: continue
+        if isinstance(node, gast.AST):
+            # In two case, the ast.Name should be filtered.
+            # 1. Function name like `my_func` of my_func(x)
+            # 2. api prefix like `fluid` of `fluid.layers.mean`
+            if isinstance(node, gast.Return):
+                continue
+            elif isinstance(node, gast.Call) and isinstance(node.func,
+                                                            gast.Name):
+                not_name_set.add(node.func.id)
+            elif isinstance(node, gast.Attribute) and isinstance(node.value,
+                                                                 gast.Name):
+                not_name_set.add(node.value.id)
+            if isinstance(
+                    node, gast.Name
+            ) and node.id not in name_ids and node.id not in not_name_set:
+                if isinstance(node.ctx, (gast.Store, gast.Load, gast.Param)):
+                    name_ids[node.id].append(node.ctx)
+            else:
+                if isinstance(node, gast.Assign):
+                    node = copy.copy(node)
+                    node._fields = ('value', 'targets')
+                for field, value in gast.iter_fields(node):
+                    value = value if isinstance(value, list) else [value]
+                    update(name_ids,
+                           get_name_ids(value, not_name_set, node_black_list))
+    return name_ids
+
+
+def parse_args(var_ids_dict, return_ids=None, ctx=gast.Load):
```

Comment:
Is this used for parsing args for cond? If yes, parse_cond_args may be better name because parse_args is very common in python

Reply:
Thanks, rename it into `parse_cond_args`